### PR TITLE
feat(contracts): remote-web pairing-request list/approve/deny wire shapes

### PR DESCRIPTION
**Pull request by noanflaherty**
## Summary
- Adds RemoteWebPairingRequestSummary, list response, action request, and deny response wire types for the upcoming host-side pairing approval surface
- Documents the three new loopback-only gateway routes in the module doc

Part of plan: web-pair-approve.md (PR 1 of 6)

### CODE REVIEW

**Comment by noanflaherty** [comment]
<details>
<summary>Plan papertrail: Pending web-pairing approval in settings</summary>

# Pending web-pairing approval in settings

## Overview

Today the only way to approve a remote-web pairing request (a browser sitting on `<publicBaseUrl>/assistant/pair` polling with its device code) is `vellum pair --web-approve <CODE>` on the host. This plan adds a host-side approval surface: the gateway's in-memory challenge store learns to list pending requests (ids, plaintext user codes, requester metadata) and to approve/deny by id; three new **loopback-only** gateway routes expose that; and the settings "Pair a device" card (desktop/local mode only, behind the existing `web-remote-ingress` client flag) polls the list and renders Approve/Deny.

Security model is unchanged: the new routes carry the same loopback-presence proof as the existing pairing routes (`auth: "none"` + `enforceLoopbackOnly`), so a remote paired session still cannot reach them. The one deliberate change is storing the plaintext user code in the store (today only hashes are kept) so the approval card can display it — the approver visually matches it against the code on the requesting device's screen, preserving the RFC 8628 binding.

**Out of scope (recorded follow-up, not in this plan):** a `denied` status for the remote poller. A denied challenge is simply deleted; the remote pairing page will surface it as expired (401 → "expired" state), which is acceptable for v1. No backwards-compat gating is needed anywhere — this feature is not user-facing yet.

**Executor notes:**
- Before editing gateway code, read `gateway/AGENTS.md` / `gateway/CLAUDE.md`; before editing web code, read `clients/web/AGENTS.md`, `clients/web/docs/CONVENTIONS.md`, and `clients/web/docs/STYLE_GUIDE.md`.
- The pairing routes are intentionally **excluded** from the gateway OpenAPI schema (see `EXCLUDED_FROM_SCHEMA` in `gateway/src/__tests__/route-schema-guard.test.ts:197`). The new routes join that exclusion list — do NOT add OpenAPI schema entries for them.

## PR 1: Wire contracts for pairing-request list/approve/deny

### Depends on
None

### Branch
web-pair-approve/pr-1-contracts

### Title
feat(contracts): remote-web pairing-request list/approve/deny wire shapes

### Files
- packages/service-contracts/src/remote-web-pairing.ts

### Implementation steps
1. In `packages/service-contracts/src/remote-web-pairing.ts`, extend the module doc's route list (lines 6–13) with the three new gateway routes:
   - `GET  /v1/remote-web/pairing-requests` — list pending challenges (loopback-only)
   - `POST /v1/remote-web/pairing-requests/approve` — approve by request id (loopback-only)
   - `POST /v1/remote-web/pairing-requests/deny` — deny (delete) by request id (loopback-only)
2. Add the new types after `RemoteWebPairingVerificationResponse` (keep the file's existing doc-comment style — every field documented, ISO-8601 note for timestamps):
   ```ts
   /** One pending challenge as shown on a host approval surface. */
   export interface RemoteWebPairingRequestSummary {
     /** Opaque server-side id used to approve or deny this request. */
     requestId: string;
     /** The human-readable code the requesting device is displaying (e.g. "ABCD-EFGH"). */
     userCode: string;
     /** Public base URL the challenge was minted for. */
     publicBaseUrl: string;
     /** ISO-8601 instant the challenge was minted. */
     requestedAt: string;
     /** ISO-8601 instant the challenge expires. */
     expiresAt: string;
     /** Client IP the mint request arrived from (edge-forwarded when tunneled). */
     requesterIp: string;
     /** User-Agent header of the mint request, or null when absent. */
     requesterUserAgent: string | null;
   }

   /** `GET /v1/remote-web/pairing-requests` success response body (200). */
   export interface RemoteWebPairingRequestListResponse {
     requests: RemoteWebPairingRequestSummary[];
   }

   /** Request body for the pairing-request approve and deny routes. */
   export interface RemoteWebPairingRequestActionRequest {
     requestId: string;
   }

   /** `POST /v1/remote-web/pairing-requests/deny` success response body (200). */
   export interface RemoteWebPairingRequestDenyResponse {
     status: "denied";
   }
   ```
   The approve route's success body reuses the existing `RemoteWebPairingVerificationResponse` — note that in its doc comment.
3. Document in the `RemoteWebPairingRequestSummary` doc comment that the plaintext `userCode` is exposed **only** on the loopback-gated list route, and that displaying it is what lets the approver match the code against the requesting device's screen (the device-flow anti-phishing binding).

### Acceptance criteria
- `bun run typecheck` (or the package's check script) passes in `packages/service-contracts`.
- Types are exported from the same module (`@vellumai/service-contracts/remote-web-pairing`) — no new entry point.
- No behavioral code changes; types + docs only.

## PR 2: Challenge store — ids, plaintext code, requester metadata, list/approve-by-id/deny

### Depends on
PR 1

### Branch
web-pair-approve/pr-2-store

### Title
feat(gateway): pairing challenge store supports listing, approve-by-id, and deny

### Files
- gateway/src/remote-web/pairing-challenge-store.ts
- gateway/src/http/routes/remote-web-pairing-challenge.ts
- gateway/src/__tests__/remote-web-pairing-challenge.test.ts
- gateway/src/__tests__/remote-web-pairing-verification.test.ts (only if signature changes ripple into its fixtures)

### Implementation steps
1. In `gateway/src/remote-web/pairing-challenge-store.ts`, extend `PendingRemoteWebPairingChallenge` with:
   - `id: string` — minted as `randomBytes(16).toString("base64url")` at create time.
   - `userCode: string` — the display-form code (with dash), stored in plaintext. Add a short comment stating this is deliberate: the loopback-gated list route shows it to the host approver; the hash fields remain the lookup keys.
   - `createdAtMs: number`
   - `requesterIp: string`
   - `requesterUserAgent: string | null`
2. Change `createRemoteWebPairingChallenge(publicBaseUrl)` to `createRemoteWebPairingChallenge(publicBaseUrl, requester: { ip: string; userAgent: string | null })` and populate the new fields. Return shape is unchanged.
3. Extract the body of `approveRemoteWebPairingChallenge` (the status/expiry checks and the `pending → approved` transition, lines 150–168) into a private `approveChallenge(challenge, userCodeHash)` helper so approve-by-code and approve-by-id share one implementation, including the delete-on-expiry cleanup of both maps.
4. Add three exported functions:
   - `listPendingRemoteWebPairingChallenges(): RemoteWebPairingRequestSummary[]` (import the type from `@vellumai/service-contracts/remote-web-pairing`, matching how the store already returns wire shapes). Run `cleanupExpiredChallenges()` first, then map every challenge with `status === "pending"` to a summary (`requestedAt` / `expiresAt` via `new Date(...).toISOString()`), sorted newest-first by `createdAtMs`.
   - `approveRemoteWebPairingChallengeById(id: string): ApproveRemoteWebPairingChallengeResult` — linear scan of `challengesByUserCodeHash.values()` for a matching `id` (≤200 entries; no third map, so the cleanup/delete paths stay untouched). Not found → `{ status: "invalid" }`; otherwise delegate to the shared `approveChallenge` helper.
   - `denyRemoteWebPairingChallengeById(id: string): { status: "denied" } | { status: "invalid" }` — linear scan; only a challenge with `status === "pending"` may be denied (approved/exchanging/consumed → `invalid`). Deny deletes the challenge from **both** maps.
5. Update the one production caller, `handleCreateRemoteWebPairingChallenge` in `gateway/src/http/routes/remote-web-pairing-challenge.ts:142`, to pass `{ ip: clientIp, userAgent: req.headers.get("user-agent") }`. `clientIp` is already the edge-forwarded client IP for tunneled mints, which is exactly the requester we want to display.
6. Update store-touching tests for the new create signature, and add store-level tests (in `remote-web-pairing-challenge.test.ts` or alongside existing store coverage) for:
   - list returns only pending, non-expired challenges, newest first, with plaintext `userCode` matching the minted code and the requester metadata passed at create.
   - approve-by-id transitions `pending → approved` and the device-code exchange then succeeds (`claimRemoteWebPairingChallengeExchange` returns `approved`).
   - approve-by-id on an expired challenge returns `expired` and evicts it; unknown id returns `invalid`; approve-by-id on an `exchanging`/`consumed` challenge returns `invalid` (same semantics as approve-by-code).
   - deny removes the challenge (subsequent exchange poll returns `invalid`, list no longer contains it); deny of an approved challenge returns `invalid`.

### Acceptance criteria
- `bun test` in `gateway/` passes, including existing pairing challenge/verification/token tests.
- `approveRemoteWebPairingChallenge` (by code) behavior is byte-for-byte unchanged (first-writer CAS semantics, expiry eviction).
- No new Map is introduced; `resetRemoteWebPairingChallengesForTests` still fully resets state.

## PR 3: Loopback-only gateway routes for list/approve/deny

### Depends on
PR 1, PR 2

### Branch
web-pair-approve/pr-3-routes

### Title
feat(gateway): loopback-only pairing-request list/approve/deny routes

### Files
- gateway/src/http/routes/remote-web-pairing-requests.ts (new)
- gateway/src/index.ts
- gateway/src/__tests__/route-schema-guard.test.ts
- gateway/src/__tests__/remote-web-pairing-requests.test.ts (new)

### Implementation steps
1. Create `gateway/src/http/routes/remote-web-pairing-requests.ts` modeled on `remote-web-pairing-verification.ts` (same `jsonError` helper shape, same `enforceLoopbackOnly` + `readLimitedBody` usage; header comment explaining these are the host-approval routes and why they are loopback-gated — reuse the framing from `pair-device-client.ts` lines 11–15: loopback presence is the authorization, and a remote paired session must never reach them). Three handlers:
   - `handleListRemoteWebPairingRequests(req, clientIp)`: reject non-GET with 405 + `Allow: GET`; `enforceLoopbackOnly(req, clientIp, "remote-web-pairing-requests")`; return `Response.json({ requests: listPendingRemoteWebPairingChallenges() } satisfies RemoteWebPairingRequestListResponse, { headers: { "Cache-Control": "no-store" } })`.
   - `handleApproveRemoteWebPairingRequest(req, clientIp)`: POST-only; loopback guard; `readLimitedBody(req, 256)`; parse `{ requestId }` (non-empty string, else 400 `BAD_REQUEST`); call `approveRemoteWebPairingChallengeById`. `invalid` → 404 `INVALID_REQUEST_ID`; `expired` → 410 `EXPIRED_REQUEST`; success → 200 with the store's approved result (the `RemoteWebPairingVerificationResponse` shape), `Cache-Control: no-store`. No code-guess rate limiter — ids are server-minted, not guessable secrets typed by users; add a one-line comment saying so.
   - `handleDenyRemoteWebPairingRequest(req, clientIp)`: POST-only; loopback guard; same body parsing; `denyRemoteWebPairingChallengeById`. `invalid` → 404 `INVALID_REQUEST_ID`; success → 200 `{ status: "denied" } satisfies RemoteWebPairingRequestDenyResponse`.
2. Register the routes in `gateway/src/index.ts` directly after the existing `/v1/remote-web/pairing-token` entry (~line 894), same style (`auth: "none"`, handler passes `getClientIp()`):
   - `{ path: "/v1/remote-web/pairing-requests", method: "GET", auth: "none", ... }`
   - `{ path: "/v1/remote-web/pairing-requests/approve", method: "POST", auth: "none", ... }`
   - `{ path: "/v1/remote-web/pairing-requests/deny", method: "POST", auth: "none", ... }`
3. Add all three paths to `EXCLUDED_FROM_SCHEMA` in `gateway/src/__tests__/route-schema-guard.test.ts` (after line 207), each with the required comment, e.g. `// Loopback-only pairing-request approval surface — not part of the public gateway API`.
4. Create `gateway/src/__tests__/remote-web-pairing-requests.test.ts` mirroring the structure of `remote-web-pairing-verification.test.ts` (store reset in beforeEach via `resetRemoteWebPairingChallengesForTests`, fake-time via `setRemoteWebPairingChallengeNowForTests` where needed). Cover:
   - non-loopback client IP is rejected on all three routes (the loopback guard fires).
   - method guards (POST to list → 405, GET to approve/deny → 405).
   - end-to-end: mint a challenge (store call), list shows it with its `userCode`, approve by the listed `requestId`, then `claimRemoteWebPairingChallengeExchange` with the device code returns `approved`.
   - deny by id → list is empty and the exchange poll returns `invalid`.
   - unknown `requestId` → 404 on approve and deny; expired challenge → 410 on approve; malformed/oversized body → 400/413.

### Acceptance criteria
- `bun test` in `gateway/` passes, including `route-schema-guard.test.ts` (new routes excluded with comments) and the new test file.
- All three routes refuse non-loopback callers server-side; none require or read any auth token.
- No OpenAPI/schema artifacts are added or regenerated.

## PR 4: Web pairing client — list/approve/deny functions

### Depends on
PR 1

### Branch
web-pair-approve/pr-4-web-client

### Title
feat(web): pair-device client functions for pending pairing requests

### Files
- clients/web/src/domains/settings/pair-device/pair-device-client.ts
- clients/web/src/domains/settings/pair-device/pair-device-client.test.ts (new)

### Implementation steps
1. In `pair-device-client.ts`, add path constants `PAIRING_REQUESTS_PATH = "/v1/remote-web/pairing-requests"`, plus `/approve` and `/deny` variants.
2. Generalize the existing `postPairingRoute` fetch/parse/error logic so GET is supported — e.g. extract a `pairingRouteRequest<T>(url, init, signal)` core that `postPairingRoute` wraps, keeping `PairDeviceError` semantics identical. Do not attach any auth header (same loopback-proxy trust boundary as the existing routes; the file's header comment already explains it — extend that comment to mention the new request-approval routes).
3. Add three exported functions, typed with the PR 1 contracts:
   - `listPendingPairingRequests(args: { base: string; signal?: AbortSignal }): Promise<RemoteWebPairingRequestSummary[]>` — GET, returns `payload.requests` (tolerate a missing/non-array field by returning `[]`).
   - `approvePairingRequest(args: { base: string; requestId: string; signal?: AbortSignal }): Promise<void>` — POST `{ requestId } satisfies RemoteWebPairingRequestActionRequest`.
   - `denyPairingRequest(args: { base: string; requestId: string; signal?: AbortSignal }): Promise<void>` — same shape against the deny path.
4. Add `pair-device-client.test.ts` following the domain's existing test conventions (see `pair-device-card.test.tsx` / neighboring web tests for the fetch-mocking pattern): success shapes, non-OK responses throwing `PairDeviceError` with the server's `error.message`, network failure mapping, and abort passthrough.

### Acceptance criteria
- `bun run typecheck` and the web test suite pass for the touched domain.
- Existing `mintDevicePairing` behavior and its error copy are unchanged.
- New functions send no Authorization header and hit exactly the three new route paths.

## PR 5: Polling hook for pending pairing requests

### Depends on
PR 4

### Branch
web-pair-approve/pr-5-polling-hook

### Title
feat(web): use-pending-pairing-requests polling hook

### Files
- clients/web/src/domains/settings/pair-device/use-pending-pairing-requests.ts (new)
- clients/web/src/domains/settings/pair-device/use-pending-pairing-requests.test.ts (new)

### Implementation steps
1. Create `use-pending-pairing-requests.ts` exporting `usePendingPairingRequests(base: string | null)` (null → disabled; the caller passes `target?.base ?? null` so the hook is inert outside desktop/local mode). Model state-machine style on `use-pair-device.ts`.
2. Behavior:
   - While enabled, fetch the list immediately on mount and then every 5000 ms (module-level `const POLL_INTERVAL_MS = 5000`, matching the store's `POLL_INTERVAL_SECONDS`). Use an `AbortController` per fetch and clear the timer + abort on unmount/disable.
   - A successful poll replaces the list; a failed poll keeps the previous list (transient loopback-proxy hiccups shouldn't flash the UI empty) and records a non-fatal error string the card may surface.
   - Expose `approve(requestId)` and `deny(requestId)` async callbacks that call the PR 4 client functions, track a per-request in-flight action (`actingOn: { requestId: string; action: "approve" | "deny" } | null`), remove the request from the local list on success, and surface `PairDeviceError` messages on failure (a 404/410 just means the request expired or was handled elsewhere — treat it as success-shaped removal, since the CLI `--web-approve` path can race the UI).
   - Return `{ requests, actingOn, error, approve, deny }`.
3. Add hook tests (fake timers + mocked client module): initial fetch, interval refresh, disabled when `base` is null, approve removes the row and stops the in-flight marker, a 404 `PairDeviceError` on approve is treated as removal not error, poll failure retains the previous list.

### Acceptance criteria
- Web tests pass; no polling occurs when `base` is null or after unmount (assert via mock call counts with fake timers).
- Approving via the hook triggers an immediate list refresh or local removal — the row never lingers past the action's resolution.

## PR 6: Approval section in the Pair Device card

### Depends on
PR 5

### Branch
web-pair-approve/pr-6-approval-card

### Title
feat(web): approve pending web-pairing requests from settings

### Files
- clients/web/src/domains/settings/pair-device/pending-pairing-requests.tsx (new)
- clients/web/src/domains/settings/pair-device/pair-device-card.tsx
- clients/web/src/domains/settings/pair-device/pair-device-card.test.tsx

### Implementation steps
1. Create `pending-pairing-requests.tsx`: a section component taking the PR 5 hook's outputs (or taking `base` and calling the hook itself — prefer the latter so `PairDeviceCard` stays thin). Render nothing when `requests` is empty. For each request render:
   - the `userCode` prominently (monospace, same visual weight as the code the CLI prints) with copy instructing the user to confirm it matches the code shown on the requesting device before approving — this is the device-flow anti-phishing binding, keep the sentence.
   - requester context: `requesterUserAgent` (raw string, truncated) and `requesterIp`, plus relative requested time from `requestedAt`.
   - Approve and Deny buttons wired to the hook, disabled while that row's action is in flight; surface the hook's `error` inline.
2. Mount the section inside `PairDeviceCard` (below the existing mint UI). This intentionally inherits all three existing gates in `pair-device-card.tsx:28-36` — `resolvePairDeviceTarget()` (desktop/local mode with an on-machine gateway), the assistant-version support check, and the `webRemoteIngress` client flag — so the approval surface is exactly as visible as device pairing itself. Extend the card's header comment to say the card now also hosts the pending-request approval list.
3. Extend `pair-device-card.test.tsx`: section hidden when the list is empty; a pending request renders its user code and requester metadata; Approve/Deny invoke the hook actions; the whole card (including the section) stays hidden when the target is null or the flag is off.

### Acceptance criteria
- Web test suite passes; visual conventions follow `clients/web/docs/STYLE_GUIDE.md`.
- No new route, flag, or visibility logic is introduced — gating is solely the existing `PairDeviceCard` gates.
- With a pending remote mint (remote `/assistant/pair` page open through a tunnel), the settings card shows the request within one poll interval, Approve completes the remote page's pairing, and Deny makes the remote page show its expired state.

</details>

File: packages/service-contracts/src/remote-web-pairing.ts [hunk 1]
```diff
@@ -68,6 +74,48 @@ export interface RemoteWebPairingVerificationResponse {
   expiresAt: string;
 }
 
+/**
+ * One pending challenge as shown on a host approval surface.
+ *
+ * The plaintext `userCode` is exposed **only** on the loopback-gated list
+ * route: displaying it is what lets the approver match the code against the
+ * requesting device's screen — the device-flow anti-phishing binding.
```
[the remark this answers is not in the document]
Comment (noanflaherty) [reply packages/service-contracts/src/remote-web-pairing.ts]: Addressed in direct push to run-plan/web-pair-approve/pr-1

[the remark this answers is not in the document]
Comment (noanflaherty) [reply packages/service-contracts/src/remote-web-pairing.ts]: Addressed in direct push to run-plan/web-pair-approve/pr-1

**Comment by noanflaherty** [comment]
@codex review this PR again — the previous issues have been fixed in commit 4ea846a9f5e56eaee0affc8158c9b4741bdaaea7
